File: src/main/java/x/withlithum/neoware/server/commands/impl/BanCommand.java
/*
 * SPDX-FileCopyrightText: 2025-2026 WithLithum & contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

package x.withlithum.neoware.server.commands.impl;

import lombok.extern.slf4j.Slf4j;
import net.minestom.server.command.builder.arguments.Argument;
import net.minestom.server.command.builder.arguments.ArgumentType;
import net.minestom.server.utils.entity.EntityFinder;
import org.jspecify.annotations.NullMarked;
import x.withlithum.neoware.server.NeoServer;
import x.withlithum.neoware.server.commands.CommandArguments;
import x.withlithum.neoware.server.commands.CommandConditions;
import x.withlithum.neoware.server.commands.CommandSkeleton;
import x.withlithum.neoware.server.commands.Commands;
import x.withlithum.neoware.util.Eval;
import x.withlithum.neoware.util.text.BannedMessage;
import x.withlithum.neoware.util.text.Messages;

import java.time.Duration;
import java.time.Instant;

@NullMarked
@Slf4j
public final class BanCommand extends CommandSkeleton {
    private static final Argument<EntityFinder> ARGUMENT_TARGET = CommandArguments.singlePlayer("target");
    private static final Argument<Duration> ARGUMENT_EXPIRY = ArgumentType.Time("expiry")
        .setDefaultValue(Duration.ZERO);
    private static final Argument<String> ARGUMENT_REASON = ArgumentType.String("reason")
        .setDefaultValue("");

    public BanCommand() {
        super("ban");
    }

    @Override
    public void construct(NeoServer server) {
        conditionalSyntax(
            CommandConditions.AT_LEAST_SYS_OP,
            ARGUMENT_TARGET,
            ARGUMENT_EXPIRY,
            ARGUMENT_REASON, (sender, targetFinder, expiry, reason) -> {
                final var target = targetFinder.findFirstPlayer(sender);
                if (target == null) {
                    return failure(sender, Messages.ARGUMENT_PLAYER_NOT_FOUND);
                }

                final var blocklist = server.playerBlocklist();

                if (blocklist.lookup(target.getUuid()) != null) {
                    return failure(sender, message("already_banned", target.getName()));
                }

                final var info = blocklist.insert(target.getUuid(),
                    reason.isBlank() ? null : reason,
                    expiry.isZero() ? null : Instant.now().plus(expiry));

                target.kick(BannedMessage.create(info));
                log.info("{} banned {} ({}) for reason: {}",
                    Commands.nameOf(sender),
                    Commands.nameOf(target),
                    target.getUuid(),
                    Eval.eitherText(reason, "No reason specified")
                );

                return succeed(sender, message("success", target.getName()));
            });
    }
}

File: src/main/kotlin/x/withlithum/neoware/server/commands/builtin/BanCommand.kt
/*
 * SPDX-FileCopyrightText: 2025-2026 WithLithum & contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

package x.withlithum.neoware.server.commands.builtin

import io.github.oshai.kotlinlogging.KotlinLogging
import net.minestom.server.command.builder.arguments.ArgumentType
import x.withlithum.neoware.framework.server.NeoFrameworkServer
import x.withlithum.neoware.server.commands.FxCommand
import x.withlithum.neoware.server.commands.FxConditions
import x.withlithum.neoware.server.commands.NeoArguments
import x.withlithum.neoware.server.commands.extensions.makeOptional
import x.withlithum.neoware.server.commands.helpers.fail
import x.withlithum.neoware.server.commands.helpers.succeed
import x.withlithum.neoware.util.messages.BanMessage
import x.withlithum.neoware.util.messages.extensions.lc.lcArg
import x.withlithum.neoware.util.messages.extensions.nameOf
import java.time.Instant
import kotlin.time.toKotlinInstant

class BanCommand : FxCommand("ban") {
    companion object {
        private val ARGUMENT_TARGET = NeoArguments.onePlayer("target")
        private val ARGUMENT_EXPIRY = ArgumentType.Time("expiry")
            .makeOptional()
        private val ARGUMENT_REASON = ArgumentType.String("reason")
            .makeOptional()

        private val LOG = KotlinLogging.logger {  }
    }

    override fun construct(server: NeoFrameworkServer) {
        conditionalSyntax(
            FxConditions.IS_SYS_OP,
            ARGUMENT_TARGET,
            ARGUMENT_EXPIRY,
            ARGUMENT_REASON
        ) cmd@{ sender, targetFinder, expiry, reason ->
            val target = targetFinder.findFirstPlayer(sender) ?: return@cmd fail(
                sender,
                lcArg("player", "not_found")
            )
            if (server.banManager.lookup(target.uuid) != null) {
                return@cmd fail(sender, lcMe("already_banned", target.name))
            }

            val untilInstant: Instant? = if (expiry != null) {
                Instant.now().plus(expiry)
            } else {
                null
            }

            val info = server.banManager.insert(target.uuid, reason, untilInstant)
            target.kick(BanMessage.create(info))
            LOG.info { "${nameOf(sender)} banned ${target.username} (${target.uuid}) for reason: ${reason ?: "(No reason given)"}" }

            return@cmd succeed(sender, lcMe("success", target.name))
        }
    }
}